serverless/serverless
# feat: Support resource extensions

## What did you implement

The merge logic for extending resources with the existing framework using `resources.Resources` has some surprising behaviours.

This introduces a more formalized definition for extending resources, where `Metadata` and `Properties` will be shallowly merged (the values for existing keys will be replaced instead of merged), the `DependsOn` list will be added to, and the values for `CreationPolicy`, `DeletionPolicy`, `UpdatePolicy`, and `UpdateReplacePolicy` will be replaced if they exist in the extension.

Extension of other resource attributes is not supported at this time.

Related to but does not completely address the requirements in #6575, as fully complying with those requirements would be a breaking change.

Related to #7338.

## How can we verify it

1. Copy `serverless.yml` below to disk.

2. Run `serverless package`.

3. Examine `.serverless/cloudformation-template-update-stack.json`. Validate that the `SampleLogGroup` resource has `RetentionInDays` set to `"30"`.

See also the automated tests that validate the full behaviour.

<details>
<summary>serverless.yml</summary>

```yml
service: test

provider:
  name: aws
  runtime: go1.x
  stage: dev

functions:
  sample:
    handler: bin/sample
    events:
      - http:
          method: post
          path: /sample

resources:
  extensions:
    SampleLogGroup:
      Properties:
        RetentionInDays: '30'

package:
  exclude:
    - ./**
  include:
    - ./bin/**
```

</details>

## Todos


**_Is this ready for review?:_** YES
**_Is it a breaking change?:_** NO
